dartagnan/src/main/java/com/dat3m/dartagnan/wmm/relation/base/RelRMW.java

```diff
@@ -136,58 +137,61 @@ public TupleSet getMaxTupleSet(){
-                        // For RISCV and Power, addresses should match.
-                        BooleanFormula succCond = store.is(Tag.MATCHADDRESS) ? bmgr.and(isPair, sameAddress) : isPair;
-                        storeExec = bmgr.or(storeExec, succCond);
-                    }
+        BooleanFormula enc = bmgr.makeTrue();
+        BooleanFormula unpredictable = bmgr.makeFalse();
+        for(Event store : task.getProgram().getCache().getEvents(
+                FilterIntersection.get(FilterBasic.get(Tag.WRITE), FilterBasic.get(Tag.EXCL)))) {
+            checkState(store instanceof MemEvent, "non-memory event participating in '" + getName() + "'");
+            BooleanFormula storeExec = bmgr.makeFalse();
+            for(Tuple t : maxTupleSet.getBySecond(store)) {
+                MemEvent load = (MemEvent) t.getFirst();
+                BooleanFormula sameAddress = generalEqual(load.getMemAddressExpr(), ((MemEvent) store).getMemAddressExpr(), ctx);
+                // Encode if load and store form an exclusive pair
+                BooleanFormula isPair = exclPair(load, store, ctx);
+                BooleanFormula pairingCond = pairingCond(load, store, ctx);
+                // For ARMv8, the store can be executed if addresses mismatch, but behaviour is "constrained unpredictable"
+                // The implementation does not include all possible unpredictable cases: in case of address
+                // mismatch, addresses of read and write are unknown, i.e. read and write can use any address.
+                // For RISCV and Power, addresses should match.
+                if(store.is(Tag.MATCHADDRESS)) {
+                    pairingCond = bmgr.and(pairingCond, sameAddress);
+                } else {
+                    unpredictable = bmgr.or(unpredictable, bmgr.and(store.exec(), isPair, bmgr.not(sameAddress)));
                 }
-                enc = bmgr.and(enc, bmgr.implication(store.exec(), storeExec));
+                enc = bmgr.and(enc, bmgr.equivalence(isPair, pairingCond));
+                storeExec = bmgr.or(storeExec, isPair);
             }
+            enc = bmgr.and(enc, bmgr.implication(store.exec(), storeExec));
+        }
+        for(Tuple tuple : encodeTupleSet) {
+            MemEvent load = (MemEvent) tuple.getFirst();
+            MemEvent store = (MemEvent) tuple.getSecond();
+            enc = bmgr.and(enc, bmgr.equivalence(
+                    getSMTVar(tuple, ctx),
+                    minTupleSet.contains(tuple) ?
+                            getExecPair(tuple, ctx) :
+                            // Relation between exclusive load and store
+                            bmgr.and(
+                                    store.exec(),
+                                    exclPair(load, store, ctx),
+                                    generalEqual(load.getMemAddressExpr(), store.getMemAddressExpr(), ctx))));
```

Comment:
I think you can check if the store has `Tag.MATCHADDRESS` and if so, drop the condition on the addresses, since `exclPair` already implies it. It's a micro optimization though.